File: day4/allfleets0.cpp
#include <TMB.hpp>

template<class Type>
bool isNA(Type x){
  return R_IsNA(asDouble(x));
}


template<class Type>
Type objective_function<Type>::operator() ()
{
  DATA_MATRIX(N)
  DATA_MATRIX(F)
  DATA_MATRIX(M)
  DATA_IMATRIX(aux)
  DATA_IMATRIX(idx1)
  DATA_IMATRIX(idx2)   
  DATA_IVECTOR(fleetTypes)
  DATA_VECTOR(sampleTimes)    
  DATA_INTEGER(minYear)
  DATA_INTEGER(minAge)
  DATA_VECTOR(obs)
  DATA_IMATRIX(keyQ)   
  DATA_IMATRIX(keySd)
    
  int nobs=obs.size();
  
  PARAMETER_VECTOR(logQ)
  PARAMETER_VECTOR(logsd)
  PARAMETER_VECTOR(missing)
  vector<Type> sd=exp(logsd);

  //patch missing 
  int idxmis=0; 
  for(int i=0;i<nobs;i++){
    if(isNA(obs(i))){
      obs(i)=exp(missing(idxmis++));
    }    
  }

  Type nll=0;
  
  vector<Type> logPred(nobs);
  Type Z;
  int y, a, f;
  for(int i=0; i<nobs; ++i){
    y=aux(i,0)-minYear;
    f=aux(i,1)-1;
    a=aux(i,2)-minAge;
    Z=F(y,a)+M(y,a);
    switch(fleetTypes(f)){
      case 0:
        logPred(i) = log(N(y,a))-log(Z)+log(1-exp(-Z))+log(F(y,a));
      break;
      
      case 2:
        logPred(i) = logQ(keyQ(f,a)) + log(N(y,a)) - Z*sampleTimes(f);
      break;

      default:
        std::cout<<"Error: This fleet type not implemented yet."<<std::endl;
        exit(EXIT_FAILURE);
      break;
    }
    nll += -dnorm(log(obs(i)),logPred(i),sd(keySd(f,a)),true);
  }

  REPORT(logPred)
  return nll;
}

File: day4/allfleets0.R
load("allfleets.RData")

library(TMB)
compile("allfleets0.cpp")
dyn.load(dynlib("allfleets0"))

allfleets$keyQ <- rbind(c(NA,NA,NA,NA,NA,NA,NA,NA,NA),
                        c(NA, 0, 1, 2, 3, 4, 5, 6,NA),
                        c( 7, 8, 9,10,11,12,NA,NA,NA))

allfleets$keySd <- rbind(c( 0, 0, 0, 0, 0, 0, 0, 0, 0),
                         c(NA, 1, 1, 1, 1, 1, 1, 1,NA),
                         c( 2, 2, 2, 2, 2, 2,NA,NA,NA))

par <- list()
par$logQ <- numeric(max(allfleets$keyQ, na.rm=TRUE)+1)
par$logsd <- numeric(max(allfleets$keySd, na.rm=TRUE)+1)
par$missing <- numeric(sum(is.na(allfleets$obs)))

obj <- MakeADFun(allfleets, par, random="missing", DLL="allfleets0")
fit <- nlminb(obj$par, obj$fn, obj$gr)
est <- obj$report()$logPred

